### ISSUE TITLE
Mypy does not recognize missing method

### ISSUE THREAD

**Issue by cburgdorf** [python/mypy#6990]
In the following example, mypy does not catch that `self.meh()` does not exist. Is this expected behavior because of the static `False`?

```python
class TestCase:

    def foo(self) -> bool:
        return False and self.meh()
```

The original bug in our code base was harder to spot than this but it boils down to the same reason so I'm wondering if this is expected behavior.

If I change the example to 

```python
class TestCase:

    def foo(self) -> bool:
        return True and self.meh()
```

then `mypy` does in fact spot the missing `meh` method.

>mypy_bug.py:5: error: "TestCase" has no attribute "meh"

**Comment by JelleZijlstra** [comment]
Mypy doesn't type check code it thinks is unreachable. I'm personally not convinced that's the right behavior, because it can lead to hard-to-debug issues, but that's probably not something we can easily change.

**Comment by cburgdorf** [comment]
Hey @JelleZijlstra thank you for your prompt reply!

>I'm personally not convinced that's the right behavior, because it can lead to hard-to-debug issues

Yep, that's really unfortunate. In real code, these things can be far more subtle. The root cause in our case was a missing `@property` decorator as seen below.

```python
class TestCase():
 
    @property
    def is_running(self) -> bool:
        return not self.is_stopped and self.meh()

    #@property   # <-- if we forget the @property, the missing `meh` goes unnoticed
    def is_stopped(self) -> bool:
        return False
```

If you add the `property` decorator, mypy will rightfully catch the missing `meh()` method but if you forget the property, the missing method just goes unnoticed.

**Comment by ilevkivskyi** [comment]
One of the possible ways forward is to emit an error/warning when mypy skips some branch because ti is unreachable, see https://github.com/python/mypy/issues/2395

**Comment by ilevkivskyi** [comment]
Now that we have a dedicated issue https://github.com/python/mypy/issues/7008, I think we can close this as a duplicate.
